packages/stream/test/test-map.js

```diff
@@ -0,0 +1,77 @@
+// @ts-check
+
+import 'ses';
+import './lockdown.js';
+
+import rawTest from 'ava';
+import { wrapTest } from '@endo/ses-ava';
+import { mapReader, mapWriter, makePipe } from '../index.js';
+
+const test = wrapTest(rawTest);
+
+const textEncoder = new TextEncoder();
+const textDecoder = new TextDecoder();
+
+test('map reader', async (/** @type {import('ava').Assertions} */ t) => {
+  const [consumeBytesFrom, produceBytesTo] = makePipe();
+  const consumeFrom = mapReader(consumeBytesFrom, (
+    /** @type {Uint8Array} */ bytes,
+  ) => textDecoder.decode(bytes));
+  const produceTo = mapWriter(produceBytesTo, (/** @type {string} */ text) =>
+    textEncoder.encode(text),
+  );
+
+  const order = ['', 'Hello, World!', '\x00', '😇   '];
+
+  const makeProducer = async () => {
+    for (const value of order) {
+      // eslint-disable-next-line no-await-in-loop
+      const { done, value: actual } = await produceTo.next(value);
+      t.is(done, false);
+      t.is(actual, undefined);
+    }
+    const { done, value: actual } = await produceTo.return();
+    t.is(done, true);
+    t.is(actual, undefined);
+  };
+
+  const makeConsumer = async () => {
+    for (const expected of order) {
+      // eslint-disable-next-line no-await-in-loop
+      const { done, value: actual } = await consumeFrom.next(undefined);
+      t.is(done, false);
+      t.deepEqual(expected, actual);
+    }
+    const { done, value: actual } = await consumeFrom.return();
+    t.is(done, true);
+    t.is(actual, undefined);
+  };
+
+  await Promise.all([makeProducer(), makeConsumer()]);
+});
+
+test('transcoding stream terminated with cause', async (/** @type {import('ava').Assertions} */ t) => {
+  const [consumeStringsFrom, produceStringsTo] = makePipe();
+  const consumeFrom = mapReader(consumeStringsFrom, (
+    /** @type {Uint8Array} */ bytes,
+  ) => textDecoder.decode(bytes));
+  const produceTo = mapWriter(produceStringsTo, (/** @type {string} */ text) =>
+    textEncoder.encode(text),
+  );
+
+  const makeProducer = async () => {
+    await produceTo.throw(new Error('Exit early'));
+  };
+
+  const makeConsumer = async () => {
+    try {
+      for await (const _ of consumeFrom) {
+        t.fail();
```

Comment:
Can we add a comment here clarifying this would be caught by the catch block but not match the `error.message` test?

Reply:
I don’t follow. `t.fail()` doesn’t throw. This is to verify that no iterations transit before it throws.

Reply:
Oh it doesn't? Other test runners usually have their `.fail()` throw to interrupt immediately (aside from marking the test as failed).